src/utils.rs

```diff
@@ -56,6 +59,112 @@ pub fn validate_string(s: &str) -> bool {
     !s.is_empty()
 }
 
+#[cfg(test)]
+mod test {
+    #[allow(unused_imports)]
+    use super::*;
+
+    #[test]
+    fn slice_at_the_end() {
+        let s = "this is a test"; // No Unicode
+        assert_eq!(slice_utf8(s, 3, s.len()), &s[3..(s.len())]);
+    }
+
+    #[test]
+    fn slice_in_the_utf8() {
+        let s = "ϵthiϕś αβ a test"; // Unicode
+        assert_eq!(slice_utf8(s, 2, 6), "hiϕś");
+    }
+
+    #[test]
+    fn short() {
+        assert_eq!(slice_utf8("ö", 0, 1), "ö");
+    }
+
+    #[test]
+    fn arabic() {
+        let s = "من"; // entire string is unicode
+        assert_eq!(slice_utf8(s, 0, 2), s);
+    }
+
+    #[test]
+    fn empty() {
+        let s = "abcd";
+        assert_eq!(slice_utf8(s, 2, 2), &s[2..2]);
+    }
+}
+
+/// A function to handle slicing into strings which may contain unicode.
+///
+/// Unlike Python, Rust cares deeply about unicode strings and ensuring every
+/// slice into them remains valid. As slicing assumes that each Unicode
+/// character is either in or outside of a slice, a simple slice operation can
+/// panic when run on a Rust string. We manually slice a Rust string by
+/// iterating over it's character points. This avoids panic on unicode strings.
+///
+/// typical usage would be:
+///     `slice_utf8(simple_string, 3, 7)` instead of `&simple_string[3..7]``
```

Comment:
We should explicitly note that `low` and `high` are the index of the 'Unicode Scalar Value' character, not the byte index into the string like `&simple_string[3..7]` since they are not technically the same.